rpc/connection.go

```diff
@@ -609,6 +609,9 @@ func (c *Connection) doReconnect(ctx context.Context, disconnectStatus Disconnec
 			// A fatal error happened.
 			*reconnectErrPtr = err
 			// short-circuit Retry
+			if err != nil {
+				c.transport.Close()
```

Comment:
Currently the transport is only ever closed on `Shutdown`.  Previously, if `ShouldRetryOnReconnect` returns `false`, this function would return, but I think future calls using the same client would try to reconnect using the same transport.

If the transport is closed here, what will happen to code that relies on the old behavior?  It is required to replace their connection instance? And if the existing code already does that, are they not calling `Shutdown` properly before they create a new connection?